rfcs/011-dedicated-renku-git-user/011-dedicated-renku-git-user.md

```diff
@@ -0,0 +1,113 @@
+
+Lastly, with having a regular user being dedicated to the KG, there is no more impersonation
+of users and "fake" user activity that comes from the periodic checks and indexing of projects
+that is done by the KG. All this activity will now be associated with the Renkubot user rather
+than individual real users.
+
+## Design Detail
+
+### Helm Chart
+
+In the Helm chart we would add a hook that would create a user in the Gitlab deployment
+that Renku is connected to. The username and password will be saved in a secret. The KG
+would then have access to the secret and use it to authenticate with Gitlab.
+
+Alternatively the Helm chart can also accept existing user credentials. 
+
+### Gateway
+
+Either as part of this RFC or as an improvement the management of this secret / tokens for 
+the Renkubot user should be done by the Gateway and the KG will simply either ask the Gateway 
+for credentials or a Gateway's forward proxy will inject the right credentials in the right calls.
+
+This will also be helpful when we support multiple Gitlabs/Github. In this case the KG does
+not need to keep track which credentials to use for which Gitlab. The gateway can figure this
+out and inject the right credentials for the Renkubot user.
+
+### UI
+
+There are some changes needed in the UI for this. Namely:
+- To get a list of all projects for a user the UI would have to call Gitlab on behalf
+of the user. This is because the KG will now only know about projects to which the Renkubot
+user has been added. In other words, the KG will not know about any projects for which 
+KG integration is not active. It could figure this out for public Gitlab projects
+but it would be impossible to find private projects that are not in the KG.
+- Creating a project on Renku will have to include adding this Renkubot user to the project
+with `Reporter` permissions.
+- KG integration for a specific project can still be verified with the KG. In some cases
+the KG will simply not know anything at all for a project. In this case KG integration
+for this project is not enabled. Checking that the Renkubot user is in the project with the
+right permissions is not enough. It could be that the project has the Renkubot user but
+the KG has not fully finished indexing the project.
+- Enabling KG integration on projects that do not have it will have to be modified to also
+add the user to the project similarly to creating a project.
+- Disabling / removing KG integration results in removal of the project from the KG and removing
+the Renkubot user from the project.
+
+### KG
+
+Depending on when the Gateway refactoring is completed, the KG may not have to deal with the Renkubot
+credentials at all. This is the preferred option. But it may turn out that an initial implementation
+of this gives the Renkubot credentials directly to the KG at first.
+
+## Drawbacks
+
+1. The Renkubot user will have read access to many repos. If the credentials are stolen or leaked
```

Comment:
Actually, I could understand if this would be a serious concern for some of our users or admins.

Reply:
Yes, only recently we were talking about that with Eike again and we ended up with the same conclusion. We have to say we don't feel comfortable adding a single user to all the projects. Imagine the credentials get compromised. I cannot even think of an emergency plan we should follow in such a scenario (I guess we'd need a very carefully prepared plan for the incident). The thing is that if the credentials leak the person would be able to do whatever: log in, use the API and access the repos, remove the original members and so on. The list is almost indefinite. The Project Access Token solution with all the known issues is in my eyes far less problematic. The access token is known only to token-repository and cannot be even viewed by the owner of the project. You cannot use it to log in, either.

Obviously, renkulab.io it's maybe a bit less of an issue as we don't guarantee anything in a written form (as far as I know). But on envs like renku health or BIT? Would it be even easy to communicate that to the responsible people?

Also, what's maybe worth mentioning is the procedure of creation of the super renku user. It might be that we'll need to cede that to the admins with some very explicit message about the importance and power of this user. I wouldn't be surprised if there will be questions about why.

Reply:
@jachro I'm not sure I follow. The renku user wouldn't be the admin or maintainer of those projects. It would have the lowest possible read-only permissions.

Reply:
We would like to visit this topic. Personally I think the risk of credentials leaking is the same for a single user at it is for 20k users.